### PR TITLE
Refactor the ApiAlerts Client

### PR DESCRIPTION
**Pull request by rsarv3006**
https://github.com/apialerts/apialerts-go/issues/2

### CODE REVIEW

**Comment by rsarv3006** [comment]
@mononz @ToxicToast Thoughts? One improvement I thought about was passing the debugMode to then private func and putting the log statement behing that?

**Comment by rsarv3006** [comment]
I'm glad I haven't merged this, I have some code running in a lambda function and I have api alerts running on a different channel similar to this. Unfortunately the lambda doesn't wait for the channel to finish so it just kills the function without the event having been sent.

Going to modify this to expose both options publicly so the user can decide. With appropriate notice that the non blocking should be preferred.

File: main.go [hunk 1]
```diff
@@ -7,22 +7,63 @@ import (
...
 type Client struct {
-	apiKey string
+	apiKey  string
+	timeout time.Duration
 }
 
 func ApiAlertsClient() *Client {
+	timeout := 30 * time.Second
+
+	if os.Getenv("APIALERTS_TIMEOUT") != "" {
```
Comment (mononz) [inline main.go]: I don't think this timeout should be a system environment variable. Can we update the constructor to optionally accept a timeout second integer that (if not null) override a default timeout of 30sec?

In the future I'd like to consider adding something like an ApiAlertsConfig struct to the constructor where the user can provide overrides to default settings like apiKey, debugMode, timeout, logging. But that can wait for now.

File: main.go [hunk 2]
```diff
@@ -7,22 +7,63 @@ import (
...
+		apiKey:  os.Getenv("APIALERTS_API_KEY"),
+		timeout: timeout,
 	}
 }
 
 func (client *Client) SetApiKey(apiKey string) {
 	client.apiKey = apiKey
 }
 
+func (client *Client) SendAsync(message string, tags []string, link string, debugMode ...bool) {
```
Comment (mononz) [inline main.go]: We could make debugMode part of the constructor with the requested timeout change? Might be better than providing as an arg here?

File: main.go [hunk 3]
```diff
@@ -6,53 +6,70 @@ import (
...
+
+	if event.Message == "" {
 		return errors.New("message is required")
 	}
 
-	payload := map[string]interface{}{
-		"message": message,
-		"tags":    tags,
-		"link":    link,
+	if event.Channel == "" {
```
Comment (mononz) [inline main.go]: channel can be nil for sending the event to the 'default' channel

File: main.go [hunk 4]
```diff
@@ -81,5 +100,57 @@ func (client *Client) Send(message string, tags []string, link string) error {
 	default:
 		return errors.New("unknown error")
 	}
+}
+
+func (client *APIAlertsClient) SendWithApiKey(apiKey string, event model.APIAlertsEvent) error {
+	url := "https://api.apialerts.com/event"
```
Comment (mononz) [inline main.go]: Might be better as a const near X_VERSION

File: main.go [hunk 5]
```diff
@@ -6,53 +6,70 @@ import (
...
-
 	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
 	if err != nil {
 		return err
 	}
 
-	req.Header.Set("Authorization", "Bearer "+client.apiKey)
+	req.Header.Set("Authorization", "Bearer "+apiKey)
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("X-Integration", "golang")
```
Comment (mononz) [inline main.go]: While we are here, lets move the 'golang' to a const too near X_VERSION

**Review by mononz — APPROVED** [review]

**Pull request merged**